lib/libnxpse050/se050/glue/wraps.c

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2020 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ *
+ * OP-TEE implementation for undefined SE050 middleware symbols
+ */
+
+#include <fsl_sss_api.h>
+#include <kernel/delay.h>
+#include <nxLog.h>
+#include <printk.h>
+#include <trace.h>
+#include <wraps.h>
+
+struct strbuf {
+	char buf[MAX_PRINT_SIZE];
+	char *ptr;
+};
+
+#define MIN(a, b)				\
+(__extension__({ __typeof__(a) _a = (a);	\
+__typeof__(b) _b = (b);				\
+_a < _b ? _a : _b; }))
+
+static unsigned long next = 1;
+
+sss_status_t sm_sleep(uint32_t ms)
+{
+	mdelay(ms);
+
+	return kStatus_SSS_Success;
+}
+
+int rand(void)
+{
+	next = next * 1103515245L + 12345;
+
+	return (unsigned int)(next / 65536L) % 32768L;
```

Comment:
I think this should use optee random generator resources.

Reply:
we cant, that would cause a catch-22 since we are using the RNG from the SE050 and this call will happen before then.

is there a way to access other RNG that the one provided by crypto_rng_read or hw_get_random_byte?

Reply:
It seems strange to me to rely on a weak RNG to access to a strong RNG.

Reply:
ight. so each request to the se050x is done over i2c following the T=1 protocol with the data buffers being encrypted on both directions (these are known as _user_ calls in the code following the NXP naming convention).

And so before we put on the wire the request for RNG to the SE050 we need to encrypt it (and later on decrypt the request)

but I agree with you...this implementation for rnd which I borrowed from somewhere (dont recall where ATM) is quite crappy; IMO OP-TEE could provides an additional call for an rng _alternate_ implementation different to the crypto one. for IMX platforms, I could just use CAAM and move on but I wanted to discuss this alternative implementation interface.

what do you think?

Reply:
Maybe this deserves a way to de-register the RNG backend and register a new one that should hopefully increase Core RNG strength.

Reply:
yes